source/Dig/base.util.scriptexpression_ng.bmx

```diff
@@ -154,28 +154,31 @@ Struct STokenGroup
 	End Method
 	
 	Method GetToken:SToken(index:Int)
+		'TODO shouldn't added be checked
 		If index < token.Length
 			Return token[index]
-		ElseIf index < dynamicToken.Length - token.Length
-			Return dynamicToken[dynamicToken.Length - token.Length]
+		ElseIf index < dynamicToken.Length + token.Length
+			Return dynamicToken[index - token.Length]
 		EndIf
 	End Method
 
 	' Identical to GetToken, except it returns a new group
 	' containing only the requested token. This is used when calling
 	' a function when you only have the name and no arguments.
 	Method GetTokenGroup:STokenGroup(index:Int)
+		'TODO shouldn't added be checked
 		If index < token.Length
 			Return New STokenGroup( token[index] )
-		ElseIf index < dynamicToken.Length - token.Length
-			Return New STokenGroup( dynamicToken[dynamicToken.Length - token.Length] )
+		ElseIf index < dynamicToken.Length + token.Length
+			Return New STokenGroup( dynamicToken[index - token.Length] )
 		EndIf
 	End Method
 
 	Method AddToken(s:SToken)
 		If added < token.Length
 			token[added] = s
-		ElseIf added < dynamicToken.Length + token.Length
+		Else
+			If added >= dynamicToken.Length + token.Length - 5 Then dynamicToken = dynamicToken[..added]
```

Comment:
Ich verstehe den Hintergrund von dynamicToken noch nicht. Klar will man wenige Größenänderungen im Array zur Laufzeit, aber wenn eine nötig ist, spielt es denn dann eine Rolle, ob token oder ein zweites Array angefasst werden müssen?
Mit zwei Arrays ist der Zugriffscode unübersichtlich (und fehleranfälliger).

Reply:
Es geht hier darum, dass static arrays GC-frei sind, dynamic arrays GC-managed. Ziel der ganzen neuen Script-Expressions war auch, den Aufwand fuer den GC niedrig zu halten (wenig neue strings, wenig dynamische arrays).

If checks sind weit billiger zu haben als arrays die GC-managed sind.

Grundgedanke ist, dass irgendwas den GC bei einigen Nutzern zum Stottern brachte (Senderkarte...) und ich da lieber das Risiko dazu minimiert sehen moechte. Auch ist natuerlich die Vermeidung des GCs oft mit Performancegewinn verbunden (Stack vs Heap).

Reply:
Fehleranfaelliger ist der Code nur, wenn du an der Interna dort rumschrauben willst, nach aussen hin ist das meines Erachtens nach ausreichend getestet um nicht mehr fehleranfaelliger zu sein als der Zugriff auf einen Array an sich (bspweise out of bounds).

Gleiche Ansaetze hab ich auch bei meinen bitmapfont-DrawBox()-Befehlen drin, damit bei "kurzen Texten") eben kein GC involviert wird.

Reply:
Mit der Aenderung: 
```
		Else
			If added >= dynamicToken.Length + token.Length - 5 Then dynamicToken = dynamicToken[..added]
```
Was soll sie exakt bewirken?

```
	Method AddToken(s:SToken)
		If added < token.Length
			token[added] = s
		ElseIf added < dynamicToken.Length + token.Length
			dynamicToken[added - token.Length] = s
		EndIf
		added :+ 1
	End Method
```
Nur mittels "Addtoken" wird "added" ueberhaupt erhoeht. 
Token 1-10 landen im ersten if (added bei 1 ist 0, und bei token 10 ist added an der Stelle 9).
Token 11 landet im elseif, da dynamictoken noch 0 und token ja 10 Elemente fasst  und Fuegen wir den 10. Token hinzu, ist der derzeitige Code fehlerbehaftet, da "dynamicToken" vorher nicht erstellt/vergroessert wird.

Was ich nicht verstehe, ist, was Dein "- 5" da im Code bewirken soll.
Es ist an der Stelle ja eher ueblich, den Array generell um ein paar Felder mehr zu vergroessern ("das wird nicht das einzige AddToken sein, da reservier ich gleich paar Slots um Resizes zu vermeiden").

```
Local dynamicArrayIndex:Int = added - token.Length
'resize dynamic array if needed
If dynamicArrayIndex >= dynamicToken.Length Then dynamicToken = dynamicToken[..dynamicArrayIndex + 5]

dynamicToken[dynamicArrayIndex] = s

```

Gleiches brauchen wir uebrigens auch in SetToken

Reply:
Ich brauchte zunächst Code, der überhaupt ein Resize des dynamiToken macht. Beim Select wurde nämlich ab einr bestimmten Zahl von cases zwar added erhöht, aber die Token waren überhaupt nicht gespeichert, was zu seltsamem Verhalten meiner Tests führte.